bagua/torch_api/contrib/fuse/optimizer.py

```diff
@@ -0,0 +1,383 @@
+import torch
+from typing import List
+import copy
+import logging
+from functools import reduce
+from bagua.torch_api.utils import check_contiguous, get_flattened_tensor
+import gorilla
+
+
+def flatten_param_and_states(optimizer: torch.optim.Optimizer):
+    _supported_params_types = [
+        "torch.cuda.FloatTensor",
+        "torch.cuda.HalfTensor",
+        "torch.FloatTensor",
+        "torch.HalfTensor",
+    ]
+
+    for group in optimizer.param_groups:
+
+        for param_type in _supported_params_types:
+            params = [param for param in group["params"] if param.type() == param_type]
+
+            weights = [p.data for p in params]
+            grads = [p.bagua_ensure_grad() for p in params]
+
+            state_tensors, state_scalars, succ = _get_state_by_name(
+                optimizer.state, params
+            )
+
+            if not succ:
+                continue
+
+            grouped_indices = calculate_mutual_groups(
```

Comment:
@NOBLES5E Should we simply flatten these tensors (since it is a user-aware operation), or flatten module based on its current underlying memory layout (see examples below)?
weights: [a, b] [c, d]
grads: [a, b] [c] [d]

-> fatten to :
weights: [a, b] [c, d]
grads: [a, b] [c, d]